libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1185,6 +1185,117 @@ string YulUtilFunctions::clearStorageStructFunction(StructType const& _type)
 	});
 }
 
+string YulUtilFunctions::copyArrayToStorage(ArrayType const& _fromType, ArrayType const& _toType)
+{
+	solAssert(
+		*_fromType.copyForLocation(_toType.location(), _toType.isPointer()).get() == dynamic_cast<ReferenceType const&>(_toType),
+		""
+	);
+	solUnimplementedAssert(!_fromType.isByteArray(), "");
+	solUnimplementedAssert(!_fromType.dataStoredIn(DataLocation::Storage), "");
+
+	string functionName = "copy_array_to_storage_from_" + _fromType.identifier() + "_to_" + _toType.identifier();
+	return m_functionCollector.createFunction(functionName, [&](){
+		Whiskers templ(R"(
+			function <functionName>(slot, value<?isFromDynamicCalldata>, len</isFromDynamicCalldata>) {
+				let length := <arrayLength>(value<?isFromDynamicCalldata>, len</isFromDynamicCalldata>)
+				<?isToDynamic>
+					<resizeArray>(slot, length)
+				</isToDynamic>
+
+				<?fromMemory>
+					let memoryOffset := <srcDataLocation>(value)
+				</fromMemory>
+
+				let elementSlot := <dstDataLocation>(slot)
+				let elementOffset := 0
+
+				for { let i := 0 } lt(i, length) {i := add(i, 1)} {
+					<?fromCalldata>
+						<?dynamicallyEncodedBase>
+							let <calldataOffset> := <accessCalldataTail>(value, add(value, mul(i, <stride>)))
+						<!dynamicallyEncodedBase>
+							let <calldataOffset> := add(value, mul(i, <stride>))
+						</dynamicallyEncodedBase>
+
+						<?isValueType>
+							let <elementValues> := <readFromCalldataOrMemory>(<calldataOffset>)
+							<updateStorageValue>(elementSlot, elementOffset, <elementValues>)
+						</isValueType>
+						<?isRefType>
+							<updateStorageValue>(elementSlot, <calldataOffset>)
+						</isRefType>
+					</fromCalldata>
+
+					<?fromMemory>
+						let <elementValues> := <readFromCalldataOrMemory>(memoryOffset)
+						<updateStorageValue>(elementSlot<?isValueType>, elementOffset</isValueType>, <elementValues>)
```

Comment:
This line can be used for all three cases: value type from calldata, reference type from calldata and from memory

Reply:
It can be used for all cases I think. Refactored a bit all around, moved the line at the end of the loop, just before updating storage slot and offset